File: mirror/tier2-ruby/showcase/type_generics.rb
# MIRROR V3: Tier 2 Ruby Showcase
# Category: Type System
# Module: type_generics.rb
# Purpose: Demonstrate generic type patterns

class Box
  attr_reader :item
  
  def initialize(item)
    @item = item
  end
  
  def extract
    @item
  end
  
  def transform(&block)
    Box.new(block.call(@item))
  end
end

class Pair
  attr_reader :key, :value
  
  def initialize(key, value)
    @key = key
    @value = value
  end
end

result = {
  box_int: Box.new(100).extract,
  box_str: Box.new("generic").extract,
  pair_ks: Pair.new("key", "string").key,
  pair_vi: Pair.new("value", 42).value,
  generic_active: true,
}

puts result.inspect

File: mirror/tier2-php/showcase/type_generics.php
<?php
// MIRROR V3: Tier 2 PHP Showcase
// Category: Type System
// Module: type_generics.php

class Box {
    private $item;
    
    public function __construct($item) {
        $this->item = $item;
    }
    
    public function extract() {
        return $this->item;
    }
    
    public function transform(callable $fn) {
        return new Box($fn($this->item));
    }
}

class Pair {
    private $key;
    private $value;
    
    public function __construct($key, $value) {
        $this->key = $key;
        $this->value = $value;
    }
    
    public function getKey() {
        return $this->key;
    }
    
    public function getValue() {
        return $this->value;
    }
}

$result = [
    'box_int' => (new Box(100))->extract(),
    'box_str' => (new Box('generic'))->extract(),
    'pair_ks' => (new Pair('key', 'string'))->getKey(),
    'pair_vi' => (new Pair('value', 42))->getValue(),
    'generic_active' => true,
];

echo json_encode($result);
?>
